File: LeetCode/Python/Maximum XOR for Each Query.py
"""
  Maximum XOR for Each Query
  URL: https://leetcode.com/problems/maximum-xor-for-each-query
  Difficulty: Medium
  Topic : Array, Bit Manipulation, Prefix Sum
"""

class Solution:
    def getMaximumXor(self, nums: List[int], maximumBit: int) -> List[int]:
        sums = [nums[0]]
        xors = [int(f'{nums[0] ^ (2**20 - 1):020b}'[-maximumBit:], 2)]
        for i in range(1,len(nums)):
            xor = sums[-1] ^ nums[i]
            sums.append(xor)
            xors.append(int(f'{xor ^ (2**20 - 1):020b}'[-maximumBit:], 2))
        return xors[::-1]

File: LeetCode/Medium/Maximum XOR for Each Query.swift
/*
  Maximum XOR for Each Query
  URL: https://leetcode.com/problems/maximum-xor-for-each-query
  Difficulty: Medium
  Topic : Array, Bit Manipulation, Prefix Sum
*/

class Solution {
    func getMaximumXor(_ nums: [Int], _ maximumBit: Int) -> [Int] {
        var r = [(1 << maximumBit) - 1]
        nums.forEach {
            let last = r.last!
            r.append(last ^ $0)
        }
        r = Array(r.reversed())
        r.popLast()
        return r
    }
}
